// @raycast/source-mapper/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// ../../../../node_modules/.pnpm/escape-string-regexp@2.0.0/node_modules/escape-string-regexp/index.js
var require_escape_string_regexp = __commonJS({
  "../../../../node_modules/.pnpm/escape-string-regexp@2.0.0/node_modules/escape-string-regexp/index.js"(exports2, module2) {
    "use strict";
    var matchOperatorsRegex = /[|\\{}()[\]^$+*?.-]/g;
    module2.exports = (string) => {
      if (typeof string !== "string") {
        throw new TypeError("Expected a string");
      }
      return string.replace(matchOperatorsRegex, "\\$&");
    };
  }
});

// ../../../../node_modules/.pnpm/stack-utils@2.0.6/node_modules/stack-utils/index.js
var require_stack_utils = __commonJS({
  "../../../../node_modules/.pnpm/stack-utils@2.0.6/node_modules/stack-utils/index.js"(exports2, module2) {
    "use strict";
    var escapeStringRegexp = require_escape_string_regexp();
    var cwd = typeof process === "object" && process && typeof process.cwd === "function" ? process.cwd() : ".";
    var natives = [].concat(
      require("module").builtinModules,
      "bootstrap_node",
      "node"
    ).map((n) => new RegExp(`(?:\\((?:node:)?${n}(?:\\.js)?:\\d+:\\d+\\)$|^\\s*at (?:node:)?${n}(?:\\.js)?:\\d+:\\d+$)`));
    natives.push(
      /\((?:node:)?internal\/[^:]+:\d+:\d+\)$/,
      /\s*at (?:node:)?internal\/[^:]+:\d+:\d+$/,
      /\/\.node-spawn-wrap-\w+-\w+\/node:\d+:\d+\)?$/
    );
    var StackUtils2 = class _StackUtils {
      constructor(opts) {
        opts = {
          ignoredPackages: [],
          ...opts
        };
        if ("internals" in opts === false) {
          opts.internals = _StackUtils.nodeInternals();
        }
        if ("cwd" in opts === false) {
          opts.cwd = cwd;
        }
        this._cwd = opts.cwd.replace(/\\/g, "/");
        this._internals = [].concat(
          opts.internals,
          ignoredPackagesRegExp(opts.ignoredPackages)
        );
        this._wrapCallSite = opts.wrapCallSite || false;
      }
      static nodeInternals() {
        return [...natives];
      }
      clean(stack, indent = 0) {
        indent = " ".repeat(indent);
        if (!Array.isArray(stack)) {
          stack = stack.split("\n");
        }
        if (!/^\s*at /.test(stack[0]) && /^\s*at /.test(stack[1])) {
          stack = stack.slice(1);
        }
        let outdent = false;
        let lastNonAtLine = null;
        const result = [];
        stack.forEach((st) => {
          st = st.replace(/\\/g, "/");
          if (this._internals.some((internal) => internal.test(st))) {
            return;
          }
          const isAtLine = /^\s*at /.test(st);
          if (outdent) {
            st = st.trimEnd().replace(/^(\s+)at /, "$1");
          } else {
            st = st.trim();
            if (isAtLine) {
              st = st.slice(3);
            }
          }
          st = st.replace(`${this._cwd}/`, "");
          if (st) {
            if (isAtLine) {
              if (lastNonAtLine) {
                result.push(lastNonAtLine);
                lastNonAtLine = null;
              }
              result.push(st);
            } else {
              outdent = true;
              lastNonAtLine = st;
            }
          }
        });
        return result.map((line) => `${indent}${line}
`).join("");
      }
      captureString(limit, fn = this.captureString) {
        if (typeof limit === "function") {
          fn = limit;
          limit = Infinity;
        }
        const { stackTraceLimit } = Error;
        if (limit) {
          Error.stackTraceLimit = limit;
        }
        const obj = {};
        Error.captureStackTrace(obj, fn);
        const { stack } = obj;
        Error.stackTraceLimit = stackTraceLimit;
        return this.clean(stack);
      }
      capture(limit, fn = this.capture) {
        if (typeof limit === "function") {
          fn = limit;
          limit = Infinity;
        }
        const { prepareStackTrace, stackTraceLimit } = Error;
        Error.prepareStackTrace = (obj2, site) => {
          if (this._wrapCallSite) {
            return site.map(this._wrapCallSite);
          }
          return site;
        };
        if (limit) {
          Error.stackTraceLimit = limit;
        }
        const obj = {};
        Error.captureStackTrace(obj, fn);
        const { stack } = obj;
        Object.assign(Error, { prepareStackTrace, stackTraceLimit });
        return stack;
      }
      at(fn = this.at) {
        const [site] = this.capture(1, fn);
        if (!site) {
          return {};
        }
        const res = {
          line: site.getLineNumber(),
          column: site.getColumnNumber()
        };
        setFile(res, site.getFileName(), this._cwd);
        if (site.isConstructor()) {
          Object.defineProperty(res, "constructor", {
            value: true,
            configurable: true
          });
        }
        if (site.isEval()) {
          res.evalOrigin = site.getEvalOrigin();
        }
        if (site.isNative()) {
          res.native = true;
        }
        let typename;
        try {
          typename = site.getTypeName();
        } catch (_) {
        }
        if (typename && typename !== "Object" && typename !== "[object Object]") {
          res.type = typename;
        }
        const fname = site.getFunctionName();
        if (fname) {
          res.function = fname;
        }
        const meth = site.getMethodName();
        if (meth && fname !== meth) {
          res.method = meth;
        }
        return res;
      }
      parseLine(line) {
        const match = line && line.match(re);
        if (!match) {
          return null;
        }
        const ctor = match[1] === "new";
        let fname = match[2];
        const evalOrigin = match[3];
        const evalFile = match[4];
        const evalLine = Number(match[5]);
        const evalCol = Number(match[6]);
        let file = match[7];
        const lnum = match[8];
        const col = match[9];
        const native = match[10] === "native";
        const closeParen = match[11] === ")";
        let method;
        const res = {};
        if (lnum) {
          res.line = Number(lnum);
        }
        if (col) {
          res.column = Number(col);
        }
        if (closeParen && file) {
          let closes = 0;
          for (let i = file.length - 1; i > 0; i--) {
            if (file.charAt(i) === ")") {
              closes++;
            } else if (file.charAt(i) === "(" && file.charAt(i - 1) === " ") {
              closes--;
              if (closes === -1 && file.charAt(i - 1) === " ") {
                const before = file.slice(0, i - 1);
                const after = file.slice(i + 1);
                file = after;
                fname += ` (${before}`;
                break;
              }
            }
          }
        }
        if (fname) {
          const methodMatch = fname.match(methodRe);
          if (methodMatch) {
            fname = methodMatch[1];
            method = methodMatch[2];
          }
        }
        setFile(res, file, this._cwd);
        if (ctor) {
          Object.defineProperty(res, "constructor", {
            value: true,
            configurable: true
          });
        }
        if (evalOrigin) {
          res.evalOrigin = evalOrigin;
          res.evalLine = evalLine;
          res.evalColumn = evalCol;
          res.evalFile = evalFile && evalFile.replace(/\\/g, "/");
        }
        if (native) {
          res.native = true;
        }
        if (fname) {
          res.function = fname;
        }
        if (method && fname !== method) {
          res.method = method;
        }
        return res;
      }
    };
    function setFile(result, filename, cwd2) {
      if (filename) {
        filename = filename.replace(/\\/g, "/");
        if (filename.startsWith(`${cwd2}/`)) {
          filename = filename.slice(cwd2.length + 1);
        }
        result.file = filename;
      }
    }
    function ignoredPackagesRegExp(ignoredPackages) {
      if (ignoredPackages.length === 0) {
        return [];
      }
      const packages = ignoredPackages.map((mod) => escapeStringRegexp(mod));
      return new RegExp(`[/\\\\]node_modules[/\\\\](?:${packages.join("|")})[/\\\\][^:]+:\\d+:\\d+`);
    }
    var re = new RegExp(
      "^(?:\\s*at )?(?:(new) )?(?:(.*?) \\()?(?:eval at ([^ ]+) \\((.+?):(\\d+):(\\d+)\\), )?(?:(.+?):(\\d+):(\\d+)|(native))(\\)?)$"
    );
    var methodRe = /^(.*?) \[as (.*?)\]$/;
    module2.exports = StackUtils2;
  }
});

// src/source-mapper/index.ts
var import_worker_threads = require("worker_threads");

// src/source-mapper/source-mapping.ts
var import_node_fs = __toESM(require("node:fs"));
var import_node_path = __toESM(require("node:path"));
var import_source_map = __toESM(require("@parcel/source-map"));
var sourceMapRegex = /data:application\/json[^,]+base64,/;
var mappingExclusions = [/^node:/];
var maxMappedStackLines = 1;
function sourceMapErrorMessage(message, firstStackItem, sourceMap) {
  const isNotAFunctionRegex = /^(?!\bundefined\b)\b(\w+)\b is not a function$/;
  if (!sourceMap || typeof firstStackItem.line === "undefined" || typeof firstStackItem.column === "undefined" || !isNotAFunctionRegex.test(message)) {
    return message;
  }
  const { line, column } = firstStackItem;
  const match = sourceMap.findClosestMapping(line, column);
  if (!match?.name) {
    return message;
  }
  return message.replace(/^\b(\w+)\b/, match.name);
}
function mapStackItem(stackItem, sourceMap, includeFileContent) {
  if (!stackItem.file || !stackItem.line || !stackItem.column) {
    return stackItem;
  }
  const mapping = sourceMap.findClosestMapping(stackItem.line, stackItem.column);
  if (!mapping) {
    return stackItem;
  }
  return {
    raw: stackItem.raw,
    file: import_node_path.default.resolve(import_node_path.default.dirname(stackItem.file), mapping.source ?? ""),
    fileContent: includeFileContent && mapping.source ? sourceMap.getSourceContent(mapping.source) : void 0,
    line: mapping.original?.line,
    column: mapping.original?.column,
    function: stackItem.function
  };
}
function readExternalSourceMap(file) {
  const sourceMapFile = `${file}.map`;
  if (!import_node_fs.default.existsSync(sourceMapFile)) {
    return null;
  }
  return import_node_fs.default.readFileSync(sourceMapFile).toString("utf8");
}
function readInlineSourceMap(file) {
  if (!import_node_fs.default.existsSync(file)) {
    return null;
  }
  const fileContents = import_node_fs.default.readFileSync(file).toString("utf8");
  if (!sourceMapRegex.test(fileContents)) {
    return null;
  }
  const base64Encoded = fileContents.slice(fileContents.lastIndexOf(",") + 1);
  const base64Decoded = Buffer.from(base64Encoded, "base64").toString("utf8");
  return base64Decoded;
}
function createSourceMap(file) {
  if (mappingExclusions.some((exclusion) => exclusion.test(file))) {
    return null;
  }
  const sourceMapContent = readInlineSourceMap(file) ?? readExternalSourceMap(file);
  if (!sourceMapContent) {
    return null;
  }
  const vlqMap = JSON.parse(sourceMapContent);
  const sourceMap = new import_source_map.default();
  sourceMap.addVLQMap(vlqMap);
  return sourceMap;
}
function performSourceMapping(message, stack, reactComponentStack) {
  if (!stack) {
    return { message, stack, reactComponentStack };
  }
  const sourceMaps = /* @__PURE__ */ new Map();
  let numMappedItems = 0;
  const mappedStack = stack.map((stackItem) => {
    if (numMappedItems >= maxMappedStackLines || !stackItem.file) {
      return stackItem;
    }
    try {
      const sourceMap = sourceMaps.get(stackItem.file) ?? createSourceMap(stackItem.file);
      if (!sourceMap) {
        return stackItem;
      }
      sourceMaps.set(stackItem.file, sourceMap);
      return mapStackItem(stackItem, sourceMap, numMappedItems === 0);
    } catch {
    } finally {
      numMappedItems += 1;
    }
    return stackItem;
  });
  numMappedItems = 0;
  const mappedComponentStack = reactComponentStack?.map((stackItem) => {
    if (numMappedItems >= maxMappedStackLines || !stackItem.file) {
      return stackItem;
    }
    try {
      const sourceMap = sourceMaps.get(stackItem.file) ?? createSourceMap(stackItem.file);
      if (!sourceMap) {
        return stackItem;
      }
      sourceMaps.set(stackItem.file, sourceMap);
      return mapStackItem(stackItem, sourceMap, numMappedItems === 0);
    } catch {
    } finally {
      numMappedItems += 1;
    }
    return stackItem;
  });
  const firstStackItem = stack[0];
  if (!firstStackItem || typeof firstStackItem.file === "undefined" || typeof message === "undefined") {
    return { message, stack: mappedStack, reactComponentStack: mappedComponentStack };
  }
  return {
    message: sourceMapErrorMessage(message, firstStackItem, sourceMaps.get(firstStackItem.file)),
    stack: mappedStack,
    reactComponentStack: mappedComponentStack
  };
}

// src/source-mapper/stacktrace.ts
var import_stack_utils = __toESM(require_stack_utils());
var stackExclusions = [/server\/index.js$/, /^node:internal\/process\/task_queues/];
var maxStackLines = 5;
var stackUtils = new import_stack_utils.default({ internals: import_stack_utils.default.nodeInternals() });
function parseStackTrace(stack) {
  if (!stack) {
    return void 0;
  }
  const stackItems = stack.split("\n").slice(1, maxStackLines).map((line) => {
    const parsedLine = stackUtils.parseLine(line);
    return {
      raw: line,
      file: parsedLine?.file,
      line: parsedLine?.line,
      column: parsedLine?.column,
      function: parsedLine?.function
    };
  });
  return stackItems.filter((stackItem) => {
    return !stackExclusions.some((exclusion) => {
      return exclusion.test(stackItem.file ?? "");
    });
  });
}

// src/source-mapper/index.ts
import_worker_threads.parentPort?.postMessage(
  performSourceMapping(
    import_worker_threads.workerData?.message,
    parseStackTrace(import_worker_threads.workerData?.stack),
    parseStackTrace(import_worker_threads.workerData?.reactComponentStack)
  )
);
